app/javascript/utilities/dragAndDrop.js

```diff
@@ -0,0 +1,91 @@
+
+/**
+ * A custom Preact hook used to attach drag and drop functionality to a DOM element.
+ * @example
+ * function SomeComponent(props) {
+ *   const { setElement } = useDragAndDrop({
+ *     onDrop: someDropHandler,
+ *     onDragOver: someDragOverHandler,
+ *     onDragExit: someDragExitHandler
+ *   });
+ *
+ *  const someDomRef = useRef(null);
+ *
+ *  useEffect(() => {
+ *    if (someDomRef.current) {
+ *      setElement(someDomRef.current);
+ *    }
+ *  });
+ *
+ *  return <textarea ref={someDomRef}>I'm a text area</textarea>;
+ * };
+ *
+ * @param {object} props
+ * @param {Function} props.onDragOver The handler that runs when the dragover event is fired.
+ * @param {Function} props.onDragExit The handler that runs when the dragexit/dragleave events are fired.
+ * @param {Function} props.onDrop The handler that runs when the drop event is fired.
+ */
+export function useDragAndDrop({ onDragOver, onDragExit, onDrop }) {
+  const [element, setElement] = useState(null);
+
+  useEffect(() => {
+    if (!element) {
+      return;
+    }
+
+    const noDragAndDropHandler = (event) => event.preventDefault();
+
+    document.addEventListener('dragover', noDragAndDropHandler);
+    document.addEventListener('drop', noDragAndDropHandler);
+
+    element.addEventListener('dragover', onDragOver);
+    element.addEventListener('dragexit', onDragExit);
+    element.addEventListener('dragleave', onDragExit);
+    element.addEventListener('drop', onDrop);
+
+    return () => {
+      document.removeEventListener('dragover', noDragAndDropHandler);
+      document.removeEventListener('drop', noDragAndDropHandler);
+
+      element.removeEventListener('dragover', onDragOver);
+      element.removeEventListener('dragexit', onDragExit);
+      element.addEventListener('dragleave', onDragExit);
+      element.removeEventListener('drop', onDrop);
+    };
+  }, [element, onDragOver, onDragExit, onDrop]);
+
+  return { setElement };
+}
+
+/**
+ * Registers drag and drop events for the child element that is wrapped by this component.
+ *
+ * @example
+ * <DragAndDropZone
+ *     onDragOver={someDragOverHandler}
+ *     onDragExit={someDragExitHandler}
+ *     onDrop={someDropHandler}
+ *     >
+ *    <textarea>I'm a text area</textarea>
+ * <DragAndDropZone>
+ *
+ * @param {object} props
+ * @param {JSX.Element} props.children The React element that will register it's DOM counterpart with drag and drop events.
+ * @param {Function} props.onDragOver The handler that runs when the dragover event is fired.
+ * @param {Function} props.onDragExit The handler that runs when the dragexit/dragleave events are fired.
+ * @param {Function} props.onDrop The handler that runs when the drop event is fired.
+ */
+export function DragAndDropZone({ children, onDragOver, onDragExit, onDrop }) {
+  const { setElement } = useDragAndDrop({ onDragOver, onDragExit, onDrop });
+  const dropZoneRef = useRef(null);
+
+  if (dropZoneRef.current) {
```

Comment:
This may be a silly question, but since we're initializing the reference to `null` just above this line, can the condition be true? I guess I expected some closure that gets attached to something, but maybe I'm just confused 😃

Reply:
When the initial state is set to `null`, it set `dropZoneRef.current` to `null`. `dropZoneRef` is always defined and on subsequent rerender, the ref gets its value set to the reference to the DOM node,i.e. `dropZoneRef.current`. For more info see the ref documentation, https://reactjs.org/docs/hooks-reference.html#useref .

Reply:
I read the documentation before asking the question but in the example, they capture the ref in a closure which gets attached as an event handler in the returned JSX, which makes more intuitive sense. 

But I think I get it now, the binding of ref here happens "outside" this component via `useDragAndDrop`. So the two lines in the fragment in the doc are basically split up here.